### PR TITLE
Check kernel params, extend kernel test coverage

### PR DESCRIPTION
**Pull request by db091756**
### PR Type

- Feature
- Tests

### Description

Added `__init_check__`'s to every kernel in `scalar_valued.py`.

### How Has This Been Tested?

Added additional tests to expand coverage.

### Does this PR introduce a breaking change?

(Write your answer here.)

### Screenshots

(Write your answer here.)

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Review by rg936672 — CHANGES_REQUESTED** [review]
Main comments: add any restrictions to docstrings, and test for error messages, not just error types.

Other comments:
 - If an invalid kernel type is passed to a UniCompositeKernel, it should throw a `TypeError` rather than a `ValueError`.

File: tests/unit/test_kernels.py [hunk 1]
```diff
@@ -1976,3 +2069,8 @@ def k_x_y(x_input, y_input):
                 expected_output[x_idx, y_idx] = k_x_y(x[x_idx, :], y[y_idx, :])
 
         return _Problem(x, y, expected_output, modified_kernel)
+
+    def test_invalid_base_kernel(self):
+        """Check that an error is thrown if the base kernel is not the correct type."""
+        with pytest.raises(ValueError):
+            SteinKernel(base_kernel="base_kernel", score_function=jnp.negative)
```
[the remark this answers is not in the document]
Comment (db091756) [reply tests/unit/test_kernels.py:2076]: This makes the line too long

File: coreax/kernels/scalar_valued.py [hunk 2]
```diff
@@ -142,8 +142,9 @@ class SquaredExponentialKernel(ScalarValuedKernel):
     :math:`k(x, y) = \rho * \exp(-\frac{||x-y||^2}{2 \lambda^2})` where
     :math:`||\cdot||` is the usual :math:`L_2`-norm.
 
-    :param length_scale: Kernel smoothing/bandwidth parameter, :math:`\lambda`
-    :param output_scale: Kernel normalisation constant, :math:`\rho`
+    :param length_scale: Kernel smoothing/bandwidth parameter, :math:`\lambda`, must be
+        a positive
```
Comment (rg936672) [inline coreax/kernels/scalar_valued.py:146]:
```suggestion
        positive
```

File: coreax/kernels/scalar_valued.py [hunk 3]
```diff
@@ -250,12 +252,13 @@ class RationalQuadraticKernel(ScalarValuedKernel):
...
-    :param length_scale: Kernel smoothing/bandwidth parameter, :math:`\lambda`
-    :param output_scale: Kernel normalisation constant, :math:`\rho`
+    :param length_scale: Kernel smoothing/bandwidth parameter, :math:`\lambda`, must be
+        positive
+    :param output_scale: Kernel normalisation constant, :math:`\rho`, must be positive
     :param relative_weighting: Parameter controlling the relative weighting of
         large-scale and small-scale variations, :math:`\alpha`. As
         :math:`alpha \to \infty` the rational quadratic kernel is identical to the
-        squared exponential kernel.
+        squared exponential kernel. Must be non-negative.
```
Comment (rg936672) [inline coreax/kernels/scalar_valued.py:261]:
```suggestion
        squared exponential kernel. Must be non-negative
```

File: tests/unit/test_kernels.py [hunk 4]
```diff
@@ -823,6 +827,16 @@ def expected_divergence_x_grad_y(
...
         return expected_divergence
 
+    @pytest.mark.parametrize(
+        "parameters",
+        [(-0.1, 1), (1, -0.1)],
+        ids=["non_positive_output_scale", "non_positive_constant"],
+    )
+    def test_invalid_parameters(self, parameters: Array):
+        """Test that kernel rejects bad inputs."""
+        with pytest.raises(ValueError):
```
[the remark this answers is not in the document]
Comment (rg936672) [reply tests/unit/test_kernels.py]: This should also be done on any other `with pytest.raises` usages in this file.

File: tests/unit/test_kernels.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rg936672) [reply tests/unit/test_kernels.py:2101]:
```suggestion
            SteinKernel(
                base_kernel="base_kernel",  # pyright: ignore[reportArgumentType]
                score_function=jnp.negative,
            )
```

**Review by rg936672 — CHANGES_REQUESTED** [review]
Almost there - a couple of minor formatting changes, and some more `match=` in `with pytest.raises` blocks.

File: tests/unit/test_kernels.py [hunk 5]
```diff
@@ -280,17 +280,19 @@ def test_magic_methods(  # noqa: C901
...
+                _ = kernel**1  # pyright: ignore[reportOperatorIssue]
         elif mode == "invalid_kernel_inputs":
+            with pytest.raises(TypeError):
+                _ = AdditiveKernel(1, "string")  # pyright: ignore[reportArgumentType]
             with pytest.raises(ValueError):
-                AdditiveKernel(1, "string")  # pyright: ignore
+                _ = kernel + "string"  # pyright: ignore[reportOperatorIssue]
+            with pytest.raises(ValueError):
+                _ = kernel * "string"  # pyright: ignore[reportOperatorIssue]
+            with pytest.raises(ValueError):
```
Comment (rg936672) [inline tests/unit/test_kernels.py:294]: check error messages for all of these also

File: tests/unit/test_kernels.py [hunk 6]
```diff
@@ -1976,3 +2090,11 @@ def k_x_y(x_input, y_input):
                 expected_output[x_idx, y_idx] = k_x_y(x[x_idx, :], y[y_idx, :])
 
         return _Problem(x, y, expected_output, modified_kernel)
+
+    def test_invalid_base_kernel(self):
+        """Check that an error is thrown if the base kernel is not the correct type."""
+        with pytest.raises(TypeError):
```
Comment (rg936672) [inline tests/unit/test_kernels.py:2096]: Check error message here

**Review by rg936672 — CHANGES_REQUESTED** [review]
Just a few remaining missed `with pytest.raises` that don't check error messages, then we should be done

**Comment by db091756** [comment]
> Looks good! Approved assuming all the CI passes.
Got there in the end!

**Pull request merged**